plugins/notifications-backend/src/service/router.ts

```diff
@@ -0,0 +1,362 @@
+    } else if (read === false) {
+      await store.markUnread({ user_ref: user, ids });
+
+      if (signalService) {
+        await signalService.publish({
+          recipients: [user],
+          message: { action: 'mark_unread', notification_ids: ids },
+          channel: 'notifications',
+        });
+      }
+    }
+
+    if (saved === true) {
+      await store.markSaved({ user_ref: user, ids });
+    } else if (saved === false) {
+      await store.markUnsaved({ user_ref: user, ids });
+    }
+
+    const notifications = await store.getNotifications({ ids, user_ref: user });
+    res.status(200).send(notifications);
+  });
+
+  router.post('/notifications', async (req, res) => {
+    const { recipients, origin, payload } = req.body;
+    const notifications = [];
+    let users = [];
+
+    try {
+      await authenticateService(req);
+    } catch (e) {
+      logger.error(`Failed to authenticate notification request ${e}`);
+      res.status(401).send();
+      return;
+    }
+
+    const { title, link, description, scope } = payload;
+
+    if (!recipients || !title || !origin || !link) {
+      logger.error(`Invalid notification request received`);
+      res.status(400).send();
+      return;
+    }
+
+    let entityRef = null;
+    if (recipients.entityRef && recipients.type === 'entity') {
+      entityRef = recipients.entityRef;
+    }
+
+    try {
+      users = await getUsersForEntityRef(entityRef);
+    } catch (e) {
+      logger.error(`Failed to resolve notification receiver ${e}`);
+      res.status(400).send();
+      return;
+    }
+
+    const baseNotification: Omit<Notification, 'id' | 'userRef'> = {
+      payload: {
+        ...payload,
+        severity: payload.severity ?? 'normal',
+      },
+      origin,
+      created: new Date(),
+    };
+
+    for (const user of users) {
+      const userNotification = {
+        ...baseNotification,
+        id: uuid(),
+        userRef: user,
+      };
+      const notification = await decorateNotification(userNotification);
+
+      let existingNotification;
+      if (scope) {
+        existingNotification = await store.getExistingScopeNotification({
+          user_ref: user,
+          scope,
+          origin,
+        });
+      }
+
+      let ret = notification;
+      if (existingNotification) {
+        const restored = await store.restoreExistingNotification({
+          id: existingNotification.id,
+          notification,
+        });
+        ret = restored ?? notification;
+      } else {
+        await store.saveNotification(notification);
+      }
+
+      processorSendNotification(ret);
+      notifications.push(ret);
+    }
+
+    if (signalService) {
+      await signalService.publish({
+        recipients: entityRef === null ? null : users,
+        message: {
+          action: 'new_notification',
+          notification: { title, description, link },
+        },
+        channel: 'notifications',
+      });
+    }
+
+    res.send(notifications);
```

Comment:
General note: never use `.send` unless that is exactly what you need and be careful when using it. It easily introduces XSS vulnerabilities. In this case `.json` instead.